### PR TITLE
Customizable OpenTopped Properties

### LINKED ISSUE #7: OpenTopped logic dehardcoding for Technotypes
This is a suggestion

Allow a technotype to ignore the opentopped conditions.
The conditions are OpenToppedRangeBonus, OpenToppedDamageMultiplier, and OpenToppedWarpDistance

The new logic should work as follows:
If the Opentopped Technotype has any of these flags declared on it, it will use the value declared on the Technotype, rather than as listed in [CombatDamage]. Obviously the unit will need OpenTopped=yes, and [CombatDamage]'s values will be used by default.

However if you make new flags that override the opentopped bonuses, then i'd be down for those too!

### PR DESCRIPTION
**Pull request by Otamaa**
closes #7 
this also partially solve #225 

### Customizable OpenTopped Properties

- You can now specify settings of `OpenTopped` transport locally

```ini
[SOMETECHNO]			       ; require `TechnoType` with `OpenTopped=yes`
OpenTopped.RangeBonus=1			; integer, range bonuses for troops shooting from open topped transport
OpenTopped.DamageMultiplier=1.3       ; float , damage bonuses for troops shooting from open topped transport
OpenTopped.WarpDistance=8		; integer , temporal range bonuses for troops shooting from open topped transport

### CODE REVIEW

**Comment by secsome** [comment]
I recommend to use Ares style tag naming like the following:
```ini
OpenTopped.RangeBonus
OpenTopped.DamageMultiplier
OpenTopped.WarpDistance
```

File: src/Ext/Techno/Hooks.cpp [hunk 1]
```diff
@@ -144,10 +145,83 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+	{
+		if (auto pTransportType = pTransport->GetTechnoType())
+		{
+			if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransportType))
+			{
+				R->EAX(pExt->OpenToppedRangeBonus.Get(RulesClass::Instance->OpenToppedRangeBonus));
+
+				return 0x6F72DE;
+			}
+		}
```
Comment (Thrifinesma) [inline src/Ext/Techno/Hooks.cpp:173]:
```suggestion
		if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransport->GetTechnoType()))
		{
			R->EAX(pExt->OpenToppedRangeBonus.Get(RulesClass::Instance->OpenToppedRangeBonus));

			return 0x6F72DE;
		}
```

File: src/Ext/Techno/Hooks.cpp [hunk 2]
```diff
@@ -144,10 +145,83 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+		if (auto pTransport = pThis->Transporter)
+		{
+			if (auto pTransportType = pTransport->GetTechnoType())
+			{
+				if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransportType))
+				{
+					//it is float isnt it YRPP ? , check tomson26 YR-IDB !
+					nDamageMult = pExt->OpenToppedDamageMultiplier.Get(static_cast<float>(RulesClass::Instance->OpenToppedDamageMultiplier));
+				}
+			}
```
Comment (Thrifinesma) [inline src/Ext/Techno/Hooks.cpp:198]:
```suggestion
			if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransport->GetTechnoType()))
			{
				//it is float isnt it YRPP ? , check tomson26 YR-IDB !
				nDamageMult = pExt->OpenToppedDamageMultiplier.Get(nDamageMult);
			}
```

File: src/Ext/Techno/Hooks.cpp [hunk 3]
```diff
@@ -144,10 +145,83 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+
+DEFINE_HOOK(0x6FE43B, TechnoClass_Fire_OpenTopped_DmgMult, 0x8)
+{
+	GET(TechnoClass* const, pThis, ESI);
+
+	//replacing whole check due to `fild`
+	if (pThis->InOpenToppedTransport)
+	{
+		GET_STACK(int, nDamage, STACK_OFFS(0xB4, -0x2C));
+		auto nDamageMult = static_cast<float>(RulesClass::Instance->OpenToppedDamageMultiplier);
```
Comment (Thrifinesma) [inline src/Ext/Techno/Hooks.cpp:187]:
```suggestion
		float nDamageMult = static_cast<float>(RulesClass::Instance->OpenToppedDamageMultiplier);
```

File: docs/New-or-Enhanced-Logics.md [hunk 4]
```diff
@@ -223,6 +223,17 @@ NotHuman.RandomDeathSequence=yes  ; boolean
...
 
+### Customizable OpenTopped Properties
+
+- You can now specify settings of `OpenTopped` transport locally
+
+```ini
+[SOMETECHNO]					; require `TechnoType` with `OpenTopped=yes`
+OpenTopped.RangeBonus=1			; integer, range bonuses for troops shooting from open topped transport
+OpenTopped.DamageMultiplier=1.3  ; float , damage bonuses for troops shooting from open topped transport
+OpenTopped.WarpDistance=8		; integer , temporal range bonuses for troops shooting from open topped transport
```
Comment (Metadorius) [inline docs/New-or-Enhanced-Logics.md:234]:
```suggestion
### Customizable OpenTopped properties

- You can now override `OpenTopped` properties per TechnoType.

```ini
[SOMETECHNO]                  ; TechnoType
OpenTopped.RangeBonus=        ; integer
OpenTopped.DamageMultiplier=  ; float
OpenTopped.WarpDistance=      ; integer
```

File: src/Ext/Techno/Hooks.cpp [hunk 5]
```diff
@@ -144,10 +145,74 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+				nDamageMult = pExt->OpenToppedDamageMultiplier.Get(nDamageMult);
+			}
+		}
+
+		R->EAX(Game::F2I(nDamage * nDamageMult));
+
+		return 0x6FE45A;
+	}
+
+	return 0x6FE460;
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:200]: Better use anonymous enums here, it's not instantly obvious what those returns mean.

File: src/Ext/TechnoType/Body.cpp [hunk 6]
```diff
@@ -120,6 +120,10 @@ void TechnoTypeExt::ExtData::LoadFromINIFile(CCINIClass* const pINI)
 	// Ares 0.A
 	this->GroupAs.Read(pINI, pSection, "GroupAs");
 
+	this->OpenToppedRangeBonus.Read(exINI, pSection, "OpenTopped.RangeBonus");
+	this->OpenToppedDamageMultiplier.Read(exINI, pSection, "OpenTopped.DamageMultiplier");
+	this->OpenToppedWarpDistance.Read(exINI, pSection, "OpenTopped.WarpDistance");
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.cpp:125]: Either remove the dot in tag names or add the underscore to match the writing of the tag.

File: src/Ext/Techno/Hooks.cpp [hunk 7]
```diff
@@ -126,7 +126,8 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
 	REF_STACK(args_ReceiveDamage const, receiveDamageArgs, STACK_OFFS(0xD0, -0x4));
 
 	// Die1-Die5 sequences are offset by 10
-	#define Die(x) x + 10
+	//#define Die(x) x + 10
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:129]:
```suggestion
```

File: src/Ext/Techno/Hooks.cpp [hunk 8]
```diff
@@ -144,10 +145,74 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
 		}
 	}
 
-	#undef Die(x)
+	//#undef Die(x)
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:148]:
```suggestion
```

Don't leave commented obsolete code in pull requests.

**Review by Metadorius — CHANGES_REQUESTED** [review]
Code style is good mostly. Fix stuff pointed out in comments and add readme entry for your contribution.

File: src/Ext/Techno/Hooks.cpp [hunk 9]
```diff
@@ -144,10 +144,73 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+		}
+	}
+
+	return 0;
+}
+
+DEFINE_HOOK(0x6FE43B, TechnoClass_Fire_OpenTopped_DmgMult, 0x8)
+{
+	GET(TechnoClass* const, pThis, ESI);
+	enum{ ApplyDamageMult = 0x6FE45A , ContinueCheck = 0x6FE460 };
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:176]:
```suggestion
	enum { ApplyDamageMult = 0x6FE45A, ContinueCheck = 0x6FE460 };
	
	GET(TechnoClass* const, pThis, ESI);
```

File: src/Ext/TechnoType/Body.h [hunk 10]
```diff
@@ -113,7 +117,11 @@ class TechnoTypeExt
 			OreGathering_FramesPerDir(),
 			LaserTrailData(),
 			DestroyAnim_Random(true),
-			NotHuman_RandomDeathSequence(false)
+			NotHuman_RandomDeathSequence(false),
+			OpenTopped_RangeBonus(),
+			OpenTopped_DamageMultiplier(),
+			OpenTopped_WarpDistance()
+
 		{ }
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.h:125]:
```suggestion
		{ }
```

File: src/Ext/Techno/Hooks.cpp [hunk 11]
```diff
@@ -144,10 +144,73 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+{
+	GET(TemporalClass* const, pThis, ESI);
+
+	if (auto pTransport = pThis->Owner->Transporter)
+	{
+		if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransport->GetTechnoType()))
+		{
+			R->EDX(pExt->OpenTopped_WarpDistance.Get(RulesClass::Instance->OpenToppedWarpDistance));
+
+			return 0x71A838;
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:211]:
```suggestion
			return 0x71A838;
```

File: src/Ext/Techno/Hooks.cpp [hunk 12]
```diff
@@ -144,10 +144,73 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+			if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransport->GetTechnoType()))
+			{
+				//it is float isnt it YRPP ? , check tomson26 YR-IDB !
+				nDamageMult = pExt->OpenTopped_DamageMultiplier.Get(nDamageMult);
+			}
+		}
+
+		R->EAX(Game::F2I(nDamage * nDamageMult));
+
+		return ApplyDamageMult;
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:195]:
```suggestion
		R->EAX(Game::F2I(nDamage * nDamageMult));
		return ApplyDamageMult;
```

File: src/Ext/Techno/Hooks.cpp [hunk 13]
```diff
@@ -144,10 +144,73 @@ DEFINE_HOOK(0x518505, InfantryClass_TakeDamage_NotHuman, 0x4)
...
+{
+	GET(TechnoClass* const, pThis, ESI);
+
+	if (auto pTransport = pThis->Transporter)
+	{
+		if (auto pExt = TechnoTypeExt::ExtMap.Find(pTransport->GetTechnoType()))
+		{
+			R->EAX(pExt->OpenTopped_RangeBonus.Get(RulesClass::Instance->OpenToppedRangeBonus));
+
+			return 0x6F72DE;
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.cpp:166]:
```suggestion
			R->EAX(pExt->OpenTopped_RangeBonus.Get(RulesClass::Instance->OpenToppedRangeBonus));
			return 0x6F72DE;
```

**Review by Metadorius — APPROVED** [review]

**Pull request merged**